whats-that-typeclass-Foldable/article.md

```diff
@@ -0,0 +1,364 @@
+6
+ghci> foldl (+) 0 [1, 2, 3] -- ((0 + 1) + 2) + 3
+6
+
+-- exponentiation
+ghci> foldr (^) 2 [2, 3] -- 2 ^ (3 ^ 2)
+512
+ghci> foldl (^) 2 [2, 3] -- (2 ^ 2) ^ 3
+64
+```
+
+It doesn't matter to the addition if the folding is right-to-left or left-to-right as it's an associative operator. However, when folding with exponentiation operator, the order is significant.
+
+## Generalization of `foldr` and `foldl`
+
+Fortunately, it's possible to fold not only lists! We can implement the instance of `Foldable` whenever a data type has one type argument, in other words, when its [kind](https://wiki.haskell.org/index.php?title=Kind&oldid=62154) is `* -> *`. To figure out the kind of Haskell data type, you can write `:kind` (or simply `:k`) in GHCi. To display the type of Haskell expression, use `:type` (or `:t`).
+
+For instance, a list has one type argument — the type of the list's elements:
+
+```haskell
+ghci> :kind []
+[] :: * -> *
+
+ghci> :type [1 :: Int, 2, 3]
+[1, 2, 3] :: [Int]
+ghci> :type ['a', 'b']
+['a', 'b'] :: [Char]
+```
+
+`Maybe a` data type also has one type argument — the type of presented value:
+
+```haskell
+ghci> :kind Maybe
+Maybe :: * -> *
+
+ghci> :type Just (2 :: Int)
+Just 2 :: Maybe Int
+ghci> :type Just "abc"
+Just "abc" :: Maybe [Char]
+ghci> :type Nothing
+Nothing :: Maybe a -- can't decide what type `a` is
+```
+
+`Int` and `String` have no type arguments, and you obviously can't fold them:
+
+```haskell
+ghci> :kind Int
+Int :: *
+ghci> :type (1 :: Int)
+1 :: Int
+
+ghci> :kind String
+String :: *
+ghci> :type ""
+"" :: [Char]
+ghci> :type "ab"
+"ab" :: [Char]
+```
+
+`Either a b` data type has two type arguments corresponding to `Left` and `Right` values, so there is also no reasonable way to fold an `Either a b`. But we could define `instance Foldable (Either a)` since `Either a` has an appropriate `* -> *` kind. The same for the pair data type — `instance Foldable ((,) a)` is possible. However, such instances are not intuitive since they operate on only one of type arguments.
+
+```haskell
+ghci> :kind Either
+Either :: * -> * -> *
+
+ghci> :type Left (1 :: Int)
+Left 1 :: Either Int b -- can't decide what type argument `b` is
+ghci> :type Right 'a'
+Right 'a' :: Either a Char -- can't decide what type argument `a` is
+
+ghci> :kind Either Int
+Either Int :: * -> *
+ghci> :kind (,) Char
+(,) Char :: * -> *
+```
+
+For those data types which could have `Foldable` instance, generalized versions of `foldr` and `foldl` have the following type signatures:
+
+```haskell
+foldr :: Foldable t => (a -> b -> b) -> b -> t a -> b
+foldl :: Foldable t => (b -> a -> b) -> b -> t a -> b
+```
+
+### Some instances as an example
+
+#### `Maybe a`
+
+Let's take a look at other instances of `Foldable`. One easy-to-understand but not really useful example is `Maybe a`'s instance. Folding `Nothing` returns a base element, and reducing `Just x` with `foldr f z` (or `foldl f z`) is applying `f` to `x` and `z`:
```

Comment:
I agree that `foldl` and `foldr` are not that useful for `Maybe`, but saying this for the entire instance is not fully correct because e.g. `for` or `length` over `Maybe` might be pretty helpful occasionally. Perhaps "but not really useful" part could be just stripped.